[Lens] Switch indexpattern manually

flash1293 opened the pull request:
Adds the ability to switch between index patterns on a per layer basis. The list indicates whether the layer config is compatible, switching to an index pattern results in all non-transferable operations to be removed.

![manual_indexpatternswitch](https://user-images.githubusercontent.com/1508364/62466569-e6b0d400-b791-11e9-8d5c-d7e6f42c4dde.gif)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10832/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10878/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10993/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11001/)

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/operation_definitions/date_histogram.tsx:
```diff
@@ -79,6 +79,27 @@ export const dateHistogramOperation: OperationDefinition<DateHistogramIndexPatte
+        restrictions.calendar_interval === column.params.interval &&
+        restrictions.time_zone === column.params.timeZone
+      );
+    } else {
```
For all of the isTransferable functions, I think it's important to match field names and types across index patterns, not just names.

For date histogram specifically, you have not considered the `fixed_interval` case- I think that needs to be included

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/operation_definitions/terms.tsx:
```diff
@@ -53,6 +53,15 @@ export const termsOperation: OperationDefinition<TermsIndexPatternColumn> = {
     return [];
   },
+  isTransferable: (column, newIndexPattern) => {
+    const newField = newIndexPattern.fields.find(field => field.name === column.sourceField);
```
This isn't checking field types

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/state_helpers.ts:
```diff
@@ -158,3 +160,26 @@ export function getColumnOrder(columns: Record<string, IndexPatternColumn>): str
+): IndexPatternLayer {
+  const newColumns: IndexPatternLayer['columns'] = _.pick(layer.columns, column =>
+    isColumnTransferable(column, newIndexPattern)
+  );
```
This logic isn't going as far as I was expecting to reset any defaults that might need to be changed. For example, does the operation definition need to change when switching from rollup -> regular index?

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/layerpanel.tsx:
```diff
@@ -0,0 +1,112 @@
+  return (
+    <I18nProvider>
+      {isChooserOpen ? (
+        <EuiComboBox
```
This combobox needs some extra styling, I think it's getting shorter as I type:
<img width="352" alt="Screenshot 2019-08-06 14 50 30" src="https://user-images.githubusercontent.com/666475/62567632-94d38100-b859-11e9-8070-b19c5ad394cf.png">

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/layerpanel.tsx:
```diff
@@ -0,0 +1,112 @@
+    <I18nProvider>
+      {isChooserOpen ? (
+        <EuiComboBox
+          data-test-subj="layerIndexPatternSwitcher"
```
Prefix on the test-subj? `lns_`

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/state_helpers.test.ts:
```diff
@@ -437,4 +445,174 @@ describe('state_helpers', () => {
+      expect(updatedLayer.columns).toEqual({
+        col1: layer.columns.col1,
+      });
+    });
```
I think I commented on a couple cases that I'd like to see tests for, but to reiterate:

* Name matches, but different type
* Regular -> rollup should add params
* Rollup -> regular should remove params
* fixed and calendar intervals

flash1293 commented:
Thanks for you thorough review @wylieconlon 

>  I also noticed that the left side data panel doesn't update when I change the index pattern. This might be a nice side effect to add.

Cool idea, I added that.

> For all of the isTransferable functions, I think it's important to match field names and types across index patterns, not just names.

I even wrote a test for that case, but another behavior shadowed the check. Added to all `isTransferable` implementations.

> This combobox needs some extra styling, I think it's getting shorter as I type:

I "solved" this for now, but during UI polishing we will definitely change how this looks.

> This logic isn't going as far as I was expecting to reset any defaults that might need to be changed. For example, does the operation definition need to change when switching from rollup -> regular index?

Good point, I added a separate function `transfer` to the operation definition. It makes the transition a bit lossy (you could loose your currently configured interval), but IMHO it makes sense to hide that from the user to make the workflow as simple as possible. I took a shortcut for the fixed_interval / calendar_interval case because pulling in all that logic to compare intervals would bloat the PR a lot and we don't even support this on the UI currently. I added that as an item to our technical debt list.

Now switching to a restricted index pattern switches to the minimal interval supported by the field, switching back tries to carry over the interval but if it isn't in our limited list of supported intervals, it just picks the biggest one.

Let me know whether you think this behavior makes sense.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11263/)

chrisdavies commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/layerpanel.tsx:
```diff
@@ -0,0 +1,122 @@
+
+export function LayerPanel({ state, setState, layerId }: IndexPatternLayerPanelProps) {
+  const [isChooserOpen, setChooserOpen] = useState(false);
+  const indexPatterns = useMemo(
```
I don't think we need this useMemo. state.indexPatterns changes on each keypress when editing a field label, and on most interactions, as most interactions affect index pattern state.

I do wonder if we should flatten or normalize our state structure to reduce how much needless re-rendering occurs.

Anyway, in this case, we don't need to compute this at all unless the chooser is open. What I'd do instead is pull the entire chooser code out into its own component which can make this calculation internally. That way, it's only calculated while the chooser is open, and doesn't slow down rendering in the usual case.

flash1293 replied:
`state.indexPatterns` doesn't change on each keypress, this is just containing the available index patterns and fields. But you are right, it probably makes sense to only calculate this once the chooser is actually open. I will move that around.

chrisdavies commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/layerpanel.tsx:
```diff
@@ -0,0 +1,122 @@
+      <EuiFlexGroup justifyContent="flexEnd">
+        {isChooserOpen ? (
+          <EuiFlexItem>
+            <EuiComboBox
```
Not sure if we want to fix this kind of thing in these PRs, but the styling here is a bit weird, as the combo box and text overflows its container.

flash1293 replied:
The styling is rudimentary and not supposed to stay like that for release, it's going to get polished with the rest of the config panel.

flash1293 commented:
@chrisdavies I refactored the component, could you look over it again?
@cchaos The index pattern chooser doesn't really look nice, do we want to fix that in a separate PR?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11690/)

flash1293 commented:
Discussed offline with @cchaos Both the index pattern switch in the data panel and in the config panel should use a popover instead of an inline combobox (`EuiSelectable`). This will be done in a separate PR together with unifying both dropdowns in code.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11745/)
